### PR TITLE
Generate configuration table

### PR DESCRIPTION
**Pull request by kochman**
Set up Travis so that it can run `docgen` to update the Markdown table of Sigma's configuration options and push any changes to the gh-pages branch. 

### Change log

- Create `docgen` which recurses through Sigma's global config to discover all options and spit out a Markdown table
- Add missing tags in Config structs ("desc" and "default")

### CODE REVIEW

File: docgen/deploy.sh [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+SHA=`git rev-parse --verify HEAD`
+
+# Clone the existing gh-pages for this repo into gh-pages/
+# Create a new empty branch if gh-pages doesn't exist yet (should only happen on first deploy)
+git clone --branch $TARGET_BRANCH --single-branch $REPO gh-pages
+cd gh-pages
+#git checkout $TARGET_BRANCH || git checkout --orphan $TARGET_BRANCH
+
+# Clean out existing contents
+#rm -rf gh-pages/**/* || exit 0
```
Comment (meganokeefe) [inline docgen/deploy.sh]: keep this, or remove?

File: docgen/deploy.sh [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+# Now let's go have some fun with the cloned repo
+git config user.name "Travis CI"
+git config user.email "$COMMIT_AUTHOR_EMAIL"
+
+# Run our compile script
+../docgen/generate.sh
+
+# If there are no changes to the output (e.g. this is a README update) then just bail.
+if git diff --quiet; then
+    echo "No changes to the output on this push; exiting."
```
Comment (meganokeefe) [inline docgen/deploy.sh]: NIT, can this message be slightly more specific?

File: docgen/deploy.sh [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+ENCRYPTED_KEY=${!ENCRYPTED_KEY_VAR}
+ENCRYPTED_IV=${!ENCRYPTED_IV_VAR}
+openssl aes-256-cbc -K $ENCRYPTED_KEY -iv $ENCRYPTED_IV -in ../docgen/travis-deploy.pem.enc -out ../docgen/travis-deploy.pem -d
+chmod 600 ../docgen/travis-deploy.pem
+eval `ssh-agent -s`
+ssh-add ../docgen/travis-deploy.pem
+
+# Pull requests and commits to other branches shouldn't try to deploy, just build to verify
+if [ "$TRAVIS_PULL_REQUEST" != "false" -o "$TRAVIS_BRANCH" != "$SOURCE_BRANCH" ]; then
+    echo "Skipping deploy; just doing a build."
```
Comment (meganokeefe) [inline docgen/deploy.sh]: again, slightly more specific message here?

File: docgen/generator/docgen.go [hunk 4]
```diff
@@ -0,0 +1,169 @@
+package generator
```
Comment (meganokeefe) [inline docgen/generator/docgen.go]: dumb Q, did you write this raw or pull it from somewhere?

File: docgen/deploy.sh (hunk 2 above)
Comment (kochman) [inline docgen/deploy.sh]: Yep!

File: docgen/deploy.sh (hunk 3 above)
Comment (kochman) [inline docgen/deploy.sh]: Also yep.

File: docgen/generator/docgen.go (hunk 4 above)
Comment (kochman) [inline docgen/generator/docgen.go]: Wrote it!

Comment (kochman) [inline docgen/generator/docgen.go]: Having it in its own package is probably overkill. Originally I had a separate thing that turned any struct into a slice of Fields describing its fields, and then generator took that in and spat out the table. But then I found the structs package so I got rid of that code.

File: docgen/generator/docgen.go [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+func underscore(s string) string {
+	return splitCamel(s, '_')
+}
+
+// Hyphen converts "CamelCasedString" to "camel-cased-string".
+func hyphen(s string) string {
+	return splitCamel(s, '-')
+}
+
+func splitCamel(s string, sep byte) string {
```
Comment (meganokeefe) [inline docgen/generator/docgen.go]: A++

**Comment by meganokeefe** [comment]
LGTM!

File: docgen/generator/docgen.go (hunk 5 above)
Comment (kochman) [inline docgen/generator/docgen.go]: That was totally Adam (or whoever wrote it in `config/config.go`).

**Pull request merged**